src/hooks/usePrivateChats.ts

```diff
@@ -0,0 +1,184 @@
+
+import { sendPrivateMessage, setChatMessageIsRead } from "api/chat";
+
+import { myPrivateChatsSelector } from "utils/selectors";
+import {
+  chatSort,
+  buildMessage,
+  getPreviewChatMessageToDisplay,
+  getMessageToDisplay,
+} from "utils/chat";
+import { WithId } from "utils/id";
+
+import { PreviewChatMessage, PrivateChatMessage } from "types/chat";
+
+import { isLoaded, useFirestoreConnect } from "./useFirestoreConnect";
+import { useSelector } from "./useSelector";
+import { useUser } from "./useUser";
+import { useWorldUsersById } from "./users";
+
+export type ChatToDisplay = {};
+
+export const useMyPrivateChatsConnect = () => {
+  const { user } = useUser();
+
+  useFirestoreConnect(() => {
+    if (!user?.uid) return [];
+
+    return [
+      {
+        collection: "privatechats",
+        doc: user.uid,
+        subcollections: [{ collection: "chats" }],
+        storeAs: "myPrivateChats",
+      },
+    ];
+  });
+};
+
+export const useMyPrivateMessages = () => {
+  useMyPrivateChatsConnect();
+
+  const myPrivateMessages = useSelector(myPrivateChatsSelector);
+
+  return useMemo(
+    () => ({
+      myPrivateMessages: myPrivateMessages ?? [],
+      isUserPrivateChatsLoaded: isLoaded(myPrivateMessages),
+    }),
+    [myPrivateMessages]
+  );
+};
+
+export const usePrivateChatList = () => {
+  const { user } = useUser();
+  const { worldUsersById } = useWorldUsersById();
+  const { myPrivateMessages } = useMyPrivateMessages();
+
+  const userId = user?.uid;
+
+  const recentPreviewChatMessagesMap = useMemo(
+    () =>
+      myPrivateMessages.reduce<{
+        [key: string]: PreviewChatMessage;
+      }>((acc, message) => {
+        if (!userId) return acc;
+
+        const { from: fromUserId, to: toUserId } = message;
+
+        // NOTE: Either `from` author or `to` author is Me. Filter me out
+        const counterPartyUserId =
+          fromUserId === userId ? toUserId : fromUserId;
+
+        // NOTE: Filter out not existent users
+        if (!worldUsersById[counterPartyUserId]) return acc;
+
+        if (counterPartyUserId in acc) {
+          const previousMessage = acc[counterPartyUserId];
+
+          // NOTE: If the message is older, replace it with the more recent one
+          if (previousMessage.ts_utc > message.ts_utc) return acc;
+
+          return {
+            ...acc,
+            [counterPartyUserId]: { ...message, counterPartyUserId },
+          };
+        }
+
+        return {
+          ...acc,
+          [counterPartyUserId]: { ...message, counterPartyUserId },
+        };
+      }, {}),
+    [myPrivateMessages, userId, worldUsersById]
+  );
+
+  return useMemo(
+    () => ({
+      privateChatList: Object.values(recentPreviewChatMessagesMap)
+        .sort(chatSort)
+        .map((message) =>
+          getPreviewChatMessageToDisplay(message, worldUsersById, userId)
+        ),
+    }),
+    [recentPreviewChatMessagesMap, worldUsersById, userId]
+  );
```

Comment:
Nope, I would argue that since it would be hard to add new returning options in the future. In this very case, I also return isLoaded from here(even though I don't use it).

Reply:
If we're also returning `isLoaded` now then that's a good reason to keep this returning an object.

If a hook is only returning a single item though, even if one day we might make it return another, that's not a 'good enough' reason to need to return an object IMO. The 'hardness' to refactor it later is a pretty trivial change to make, even if it touches a bunch of places.

Since you're returning this isLoaded here now as well though, all good.

Reply:
Sounds reasonable